src/DotNetWorker/Hosting/BootstrapHostBuilder.cs

```diff
@@ -0,0 +1,170 @@
+                    break;
+                }
+            }
+
+            if (Context is null)
+            {
+                throw new InvalidOperationException($"{nameof(HostBuilderContext)} must exist in the {nameof(IServiceCollection)}");
+            }
+        }
+
+        public IDictionary<object, object> Properties => Context.Properties;
+
+        public HostBuilderContext Context { get; }
+
+        public IHostBuilder ConfigureAppConfiguration(Action<HostBuilderContext, IConfigurationBuilder> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(Context, _builder.Configuration);
+            }
+            else
+            {
+                _configureAppActions.Add(configureDelegate);
+            }
+
+            return this;
+        }
+
+        public IHostBuilder ConfigureHostConfiguration(Action<IConfigurationBuilder> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(_builder.Configuration);
+            }
+            else
+            {
+                _configureHostActions.Add(configureDelegate);
+            }
+
+            return this;
+        }
+
+        public IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollection> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(Context, _builder.Services);
+            }
+            else
+            {
+                _configureServicesActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+            }
+
+            return this;
+        }
+
+        public IHost Build()
+        {
+            // Functions configuration will never call this.
+            throw new InvalidOperationException();
+        }
+
+        public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderContext, TContainerBuilder> configureDelegate)
+        {
+            // Functions configuration will never call this.
+            throw new InvalidOperationException();
```

Comment:
nit: Do we want to add specific messages to these exceptions in case they are ever called somehow?

Reply:
I don't think that's necessary. These are all only accessible internally so debugging them should be fairly easy with stack trace. And if we do hit them, it'd likely be during dev work.